ocaml/xapi/xapi_database_backup.ml

```diff
@@ -0,0 +1,175 @@
+  req.Http.Request.close <- true;
+  let token = Generation.of_string (snd (List.find (fun (x,y) -> x = "token") req.query)) in
+  Xapi_http.with_context "Database backup" req s (fun __context ->
+      let d = get_delta __context token in
+      let str_update = (Jsonrpc.to_string (rpc_of_delta d)) in
+      Http_svr.response_str req s str_update
+    )
+
+(* Slave Functions *)
+
+let send_notification (c:Context.t) reason p =
+  let record = (Eventgen.find_get_record p.tbl ~__context:c ~self:p.row) () in
+  begin match record with
+    | None -> debug "Could not send %s event for %s in %s table" reason p.row p.tbl;
+    | Some record -> debug "Sending %s event for %s object in %s table" reason p.row p.tbl;
+      Db_action_helper.events_notify ~snapshot:record p.tbl reason p.row;
+  end
+
+let apply_deletes __context (deletes:del_tables list) =
+  List.iter
+    (fun dtbl ->
+       (fun tblname objref ->
+          if Db_cache_impl.is_valid_ref (Context.database_of __context) objref then
+            Xapi_slave_db.slave_db := Db_cache_types.remove_row tblname objref !Xapi_slave_db.slave_db
+          else
+            Printf.printf "Not valid ref"
+       ) dtbl.table dtbl.key
+    ) deletes
+
+let make_change __context mtime tblname objref fldname newval =
+  Xapi_slave_db.slave_db :=
+    Database.update (
+      (fun _ -> newval)
+      |> Row.update mtime fldname newval
+      |> Table.update mtime objref Row.empty
+      |> TableSet.update mtime tblname Table.empty) !Xapi_slave_db.slave_db
+
+let unpack_row __context (table:table) tblname row =
+  List.iter (fun (f:field)->
+      Mutex.execute Xapi_slave_db.slave_db_mutex
+        (fun () -> make_change __context f.stat.modified tblname row.objref f.fldname f.value)
+    ) row.value
+
+let unpack_table __context (t:table) =
+  List.iter (unpack_row __context t t.tblname) (if t.value = [] then [empty_row] else t.value)
+
+let apply_updates __context updates =
+  List.iter (unpack_table __context) updates
+
+let get_table_list db =
+  Db_cache_types.TableSet.fold_over_recent (-2L) (fun table stat value acc -> table::acc) (Db_cache_types.Database.tableset db) []
+
+let counter db table =
+  let t = Db_cache_types.TableSet.find table (Db_cache_types.Database.tableset db) in
+  {tblname=table; count=0; del=(Db_cache_types.Table.get_deleted_len t)}
+
+let object_count db =
+  let tables = get_table_list db in
+  List.map (counter db) tables
+
+let count_eq_check c1 c2 =
+  c1.tblname = c2.tblname && c1.count = c2.count && c1.del = c2.del
+
+let apply_changes (delta:delta) __context =
+  if (Db_cache_types.Manifest.generation (Db_cache_types.Database.manifest !Xapi_slave_db.slave_db)) < delta.fresh_token then
+    begin
+      Xapi_slave_db.slave_db := Database.set_generation delta.fresh_token !Xapi_slave_db.slave_db;
+      apply_updates __context delta.tables;
+      apply_deletes __context delta.deletes;
+    end;
+  (* Need to check that we haven't missed anything *)
```

Comment:
Why?

Reply:
It counts the objects in case we are in the scenario where more than the maximum length of the deleted items list have occurred. In which case the objects that we have may not be correct. Though I'm not sure what we do in this scenario is necessarily 'the right thing'.